commands.go

```
@@ -0,0 +1,746 @@
+	req.Header.Add("Raven-Client-Version", "4.0.0.0")
+	req.Header.Add("Accept", "application/json")
+	return req, nil
+}
+
+// TODO: do I need to explicitly enable compression or does the client does
+// it by default? It seems to send Accept-Encoding: gzip by default
+func simpleExecutor(n *ServerNode, cmd *RavenCommand) (*http.Response, error) {
+	req, err := makeHTTPRequest(n, cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+	rsp, err := client.Do(req)
+	// this is for network-level errors when we don't get response
+	if err != nil {
+		return nil, err
+	}
+	// we have response but it could be one of the error server response
+
+	// convert 400 Bad Request response to BadReqeustError
+	if rsp.StatusCode == http.StatusBadRequest {
+		var res BadRequestError
+		err = decodeJSONFromReader(rsp.Body, &res)
+		if err != nil {
+			return nil, err
+		}
+		return nil, &res
+	}
+
+	// convert 503 Service Unavailable to ServiceUnavailableError
+	if rsp.StatusCode == http.StatusServiceUnavailable {
+		var res ServiceUnavailableError
+		err = decodeJSONFromReader(rsp.Body, &res)
+		if err != nil {
+			return nil, err
+		}
+		return nil, &res
+	}
+
+	// convert 500 Internal Server to InternalServerError
+	if rsp.StatusCode == http.StatusInternalServerError {
+		var res InternalServerError
+		err = decodeJSONFromReader(rsp.Body, &res)
+		if err != nil {
+			return nil, err
+		}
+		return nil, &res
+	}
+
+	// convert 409 Conflict to ConflictError
+	if rsp.StatusCode == http.StatusConflict {
+		var res ConflictError
+		err = decodeJSONFromReader(rsp.Body, &res)
+		if err != nil {
+			return nil, err
+		}
+		return nil, &res
+	}
+
+	// convert 404 Not Found to NotFoundError
+	if rsp.StatusCode == http.StatusNotFound {
+		// TODO: does it ever return non-empty response?
+		res := NotFoundError{
+			URL: req.URL.String(),
+		}
+		return nil, &res
+	}
+
+	// TODO: handle other server errors
```

For unknown errors, use the following logic `2xx` is success, anything else is bad and should return an error